ministryofjustice/laa-apply-for-criminal-legal-aid
# [CRIMAPP-700] Add property ownership validation

## Description of change
Adds property ownership validation to the property and property owners screens
Property ownership needs to add up to 100 for all property owners, else a error message is shown 
Percentages entered must also be between 1 and 100 

The following scenarios are covered: 
- Displaying an error message when a percentage entered is under 1 or over 100
- Validating the percentage ownership total equals 100 if the client is the sole owner 
- Validating the percentage ownership total equals 100 if there is a partner who's percentage ownership information has been entered and no other owners 
- If a user has selected that there are other owners but has not yet entered the other owner information (as they have not yet reached the screen), the % ownership check is not undertaken 
- If a user has navigated back to the property page with incomplete property owner information, the % ownership check is not undertaken 

## Link to relevant ticket
[CRIMAPP-700](https://dsdmoj.atlassian.net/browse/CRIMAPP-700)

## Notes for reviewer

## Screenshots of changes (if applicable)

### Before changes:

### After changes:

<img width="907" alt="Screenshot 2024-04-22 at 10 58 09" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/c9f428c1-f397-4418-94ba-1118d66c9663">
<img width="907" alt="Screenshot 2024-04-22 at 10 58 21" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/58b19272-9ff3-4ab4-8fa9-94634b1f8d8c">
<img width="764" alt="Screenshot 2024-04-22 at 11 04 12" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/43149f97-ce48-485f-a56a-4f1c3bbae65a">
<img width="915" alt="Screenshot 2024-04-22 at 11 10 25" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/2b10a4f6-ee47-480e-bc26-89590a17fbf0">


https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/3cc45fea-af31-4cac-9ff9-fdd43af7d40d


## How to manually test the feature

Scenario 1: Client is sole owner

1. Start an means tested application with an either way case type
2. Select no to partner
3. Enter a nino
4. Select no to passporting benefit
5. Go through the income, outgoings and capital screens until you get to the property type page
6. Select any property type 
7. Enter a client percentage ownership over 100 and select no to other owners
8. Enter any other data for the other questions
9. Select save and continue 
10. Confirm there is an error message displayed: `The percentage of the land your client owns must be a number between 1 and 100`
11. Change the percentage ownership figure to less than 1 
12. Confirm there is an error message displayed:  `The percentage of the land your client owns must be a number between 1 and 100` 
13. Enter a percentage over 1 and less than 100 
14. Confirm there is an error message displayed: `Percentages entered need to total 100%`
15. Enter a percentage of 100 
16. Confirm you can progress to the next screen 

Scenario 2: Client and partner owns but no other owners

1. Staying on the current application, route to the partner screen and select yes to having a partner 
2. Manually route back to the property screen and either start a new property or test with the previously started property
3. You should now see a partner percentage ownership field 
4. Enter a **partner** percentage ownership over 100 and select no to other owners
5. Enter any other data for the other questions (including a valid percentage for the client percentage ownership)
6. Select save and continue 
7. Confirm there is an error message displayed: `The percentage of the land your client's partner owns must be a number between 1 and 100`
8. Change the **partner** percentage ownership figure to less than 1 
9. Confirm there is an error message displayed:  `The percentage of the land your client's partner owns must be a number between 1 and 100` 
10. Enter a percentage that does not total for both the client and partner (e.g. 60 and 20)
11. Confirm there is an error message displayed: `Percentages entered need to total 100%`
12. Enter a total percentage ownership of 100 
13. Confirm you can progress to the next screen 

Scenario 3: Client and partner own and there are other owners

1. Manually route back to the property screen and select yes for other owners 
2. Change the total percentage ownership of the client and partner to be less than 100 
3. Confirm that you can progress to the next screen when yes is selected for other owners and the percentage ownership validation does not occur on that screen
4. Fill in the information until you reach the other property owners screen 
5. Enter an additional owner with a percentage ownership over 100
6. Select save and continue 
7. Confirm there is an error message displayed: `The percentage of the property they own must be a number between 1 and 100`
8. Change the other owner percentage ownership figure to less than 1 
9. Confirm there is an error message displayed: `The percentage of the property they own must be a number between 1 and 100`
10. Enter a percentage that does not total 100 for the client and partner and other owner(s) (e.g. 60 and 20)
11. Confirm there is an error message displayed: `Percentages entered need to total 100%`
12. Enter a total percentage ownership of 100 for all owners
13. Confirm you can progress to the next screen 

Scenario 4: Client and partner own and there are other owners with incomplete details 

1. Manually route back to the other owners screen and remove some details 
2. Select save and come back later
3. Manually route to the client's assets page `steps/capital/clients_assets`
4. You should have an incomplete property
5. Click the change link and confirm that you can select save and continue and progress back to the other property owners page from the property page 
6. On the other property owners screen, click save and continue 
7. Confirm there is an error message displayed: `Percentages entered need to total 100%` and the validation occurs on this page 

[CRIMAPP-700]: https://dsdmoj.atlassian.net/browse/CRIMAPP-700?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
